### PR TITLE
community: Bedrock add support for mistral models

### PR DESCRIPTION
**Pull request by AnisZakari**
*Description**: My previous [PR](https://github.com/langchain-ai/langchain/pull/18521) was mistakenly closed, so I am reopening this one. Context: AWS released two Mistral models on Bedrock last Friday (March 1, 2024). This PR includes some code adjustments to ensure their compatibility with the Bedrock class.

### CODE REVIEW

**Comment by AnisZakari** [comment]
@3coins  @efriis

**Comment by 3coins** [comment]
@AnisZakari 
Thanks for working on these changes. Similar comment as [PR #18687](https://github.com/langchain-ai/langchain/pull/18687) about support for BedrockChat class.

**Comment by AnisZakari** [comment]
> @AnisZakari Thanks for working on these changes. Similar comment as [PR #18687](https://github.com/langchain-ai/langchain/pull/18687) about support for BedrockChat class.

Thank you @3coins for your comment, I have just pushed some code to to adapt mistral to BedrockChat class

File: libs/community/langchain_community/chat_models/mistral.py [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+        raise ValueError(f"Got unknown type {message}")
+    return message_text
+
+
+def convert_messages_to_prompt_mistral(messages: List[BaseMessage]) -> str:
+    """Convert a list of messages to a prompt for mistral."""
+
+    return "\n".join(
+        [_convert_one_message_to_text_mistral(message) for message in messages]
+    )
```
Comment (3coins) [inline libs/community/langchain_community/chat_models/mistral.py:31]: I do not recommend adding anything in the `mistral.py` files. Rather, just copy these to the bedrock code.
```suggestion

```

**Review by 3coins — CHANGES_REQUESTED** [review]
@AnisZakari 
See my comment about moving code out of `mistral.py`, plz delete that file. Also, it seems like the `async` streaming support is missing for mistral here, can you bring in code from PR [#18687](https://github.com/langchain-ai/langchain/pull/18687) and see if that works.

File: libs/community/langchain_community/chat_models/mistral.py (hunk 1 above)
Comment (AnisZakari) [reply libs/community/langchain_community/chat_models/mistral.py]: thanks @3coins !
I deleted `mistral.py` and moved the code to `bedrock.py `
I've also added the code for async streaming support

I've tested and everything seems to work :)

**Review by 3coins — APPROVED** [review]
@AnisZakari 
Awesome job! Thanks for working on this.

**Comment by 3coins** [comment]
@efriis 
Can you please help merge this change. Thanks.

**Review by efriis — APPROVED** [review]

**Comment by Maxinho96** [comment]
Hi @AnisZakari, by looking at the code looks like the output format will be the following:
```
<<SYS>> System message <</SYS>>
[INST] Human message [/INST]
AI message
```
But to my understanding, the format should be the following instead:
```
<s>[INST] System/Human message [/INST] AI message</s>[INST] Human message [/INST]
```
The `<<SYS>>` tag is not used by Mistral, just `[INST]` for both System and Human messages. And `</s>` closing tag should be used after the AI message. Also, looks like newlines are not needed.

References:

- Bedrock docs: https://docs.aws.amazon.com/bedrock/latest/userguide/model-parameters-mistral.html#model-parameters-mistral-request-response
- Mistral docs: https://docs.mistral.ai/models/#chat-template

**Pull request merged**